Changes for file based catalog in Makefile and DEVELOPMENT.md

tbuskey opened the pull request:
**- Description of the problem which is fixed/What is the use case**
[KATA-3122](https://issues.redhat.com/browse/KATA-3122) Changes for FBC catalog to replace sqlite

**- What I did**
- Makefile
  - add an token variable for docker-build so the public images are not used
  - change the (downloaded) opm version to 1.28.0
  - change catalog-build for file based catalogs, replacing sqlite
- docs/DEVELOPMENT.md
  - explain how to use the token instead of public images
  - sed to change SANDBOXED_CONTAINERS_EXTENSION for openshift
  - where to add SKIP_TESTS

**- How to verify it**
Build per DEVELOPMENT.md
Create a catsrc using your catalog & install in your cluster
Install the operator in your cluster

**- Description for the changelog**
Changes for file based catalog instead of sqlite

tbuskey commented:
/assign @gkurz 
/assign @littlejawa 
/assign @cpmeadors

gkurz commented:
Hi @tbuskey !

This is conflating way to many changes into a single commit. Please break it down into smaller items.
First starting with preparatory patches and then adding the actual support for file based catalog.

- add authentication support
- automate download of operator-sdk
- bump the version of opm
- file based catalog

You can regroup all the documentation changes in a final commit, for simplicity.

gkurz commented on Makefile:
```diff
@@ -143,7 +145,7 @@ run: manifests generate fmt vet ## Run a controller from your host.
 docker-build: test ## Build docker image with the manager.
 	docker build \
-		-t ${IMG} \
+		-t ${IMG} $(IMAGE_AUTH) \
```
The meaning of IMAGE_AUTH is is a bit unclear. It exposes a direct dependency to the docker command line for no good reason. I'd prefer IMAGE_AUTH to be called something like AUTH_FILE and do
```suggestion
		-t ${IMG} \
		--authfile $(AUTH_FILE) \
```

tbuskey replied:
Changed to AUTH_FILE 
Explanation is in DEVELOPMENT.md

bpradipt commented on Makefile:
```diff
@@ -246,6 +248,23 @@ bundle-build: bundle ## Build the bundle image.
 bundle-push: ## Push the bundle image.
 	$(MAKE) docker-push IMG=$(BUNDLE_IMG)
 
+.PHONY: operator-sdk
```
I personally prefer to keep the dev prerequisites (golang, operator-sdk etc) outside of the Makefile, since there could be different ways to install and configure it. For example operator-sdk might be on my system path for usage across different operators.
Instead, checking for the presence of specific version of operator-sdk binary (based on osc operator requirement) in the Makefile and throwing error is portable and useful imho.

tbuskey replied:
Sounds reasonable.

Does this work?
1. check if operator-sdk already exists in the path
    - download into ./bin IFF it doesn't
 2.  check the version of operator-sdk against $OPERATOR-SDK-VERSION (allow overrides)
     - fail make if version isn't matched

tbuskey replied:
I'm leaving the operator-sdk section unchanged

bpradipt commented on docs/DEVELOPMENT.md:
````diff
@@ -2,30 +2,35 @@
+#### Using the token
+Set the authfile used by docker-build if login was successful.  Ex:
+```
+export IMAGE_AUTH='--authfile=~/.docker/config.json'
````
Is this really needed? Doesn't docker honour the default auth file (`$HOME/.docker/config.json`) automatically if present ?

tbuskey replied:
`docker-build` did not use the auth I had in `$HOME/.docker/config.json` until I added
 `--authfile=$HOME/.docker/config.json`

If the user has not set IMAGE_AUTH, the Makefile will leave it out

tbuskey commented:
/retest

tbuskey commented:
@gkurz @bpradipt 
Changes are now in 5 commits
There are no changes for `operator-sdk`

tbuskey commented:
/retest

cpmeadors commented on Makefile:
```diff
@@ -270,17 +272,25 @@ BUNDLE_IMGS ?= $(BUNDLE_IMG)
+catalog-build: opm
+	@[ ! -d catalog ] && mkdir -p catalog || true
+	@[ ! -f catalog.Dockerfile ] && $(OPM) generate dockerfile catalog || true
+	$(OPM) init  sandboxed-containers-operator --default-channel=$(DEFAULT_CHANNEL) --description ./README.md --output yaml > catalog/index.yaml
```
This should be done once and then checked in to the repo.  We need static files that we can modify as needed to make thing clear.

gkurz replied:
If we are to check in generated files, then I'd rather not have them created as part of the main build flow.

tbuskey replied:
catalog-init is not in the main build flow.  Normal is make catalog-build; make catalog-push

If you are using a private quay.io, the catalog files need to be re-written.  make catalog-clean; make catalog-init will do that for you.

cpmeadors commented on Makefile:
```diff
@@ -310,7 +320,13 @@ bundle-clean: ## Clean generated bundle files
 
+.PHONY: catalog-clean
+catalog-clean: ## Clean catalog files
+	$(RM) -r catalog
```
This needs to be removed once the catalog dir has been checked in to the repo.

gkurz replied:
Both `catalog` initial creation and removal should be dropped from the rules in this case.

tbuskey replied:
removing catalog-init/catalog-clean from Makefile
Adding directions to generate catalog to DEVELOPMENT.md

tbuskey replied:
Dropped

cpmeadors commented:
Probably need to add stuff to .gitignore as describe at https://olm.operatorframework.io/docs/reference/file-based-catalogs/#structure

tbuskey commented:
> Probably need to add stuff to .gitignore as describe at https://olm.operatorframework.io/docs/reference/file-based-catalogs/#structure

bundle/ is

tbuskey commented:
@gkurz @bpradipt could you approve?

gkurz commented:
> @gkurz @bpradipt could you approve?

Sorry for coming back late @tbuskey. Thanks for splitting the patches !

Some more general polishing is needed though :
- provide a description in every commit (e.g. first commit adds authentication but we don't know why)
- add you Signed-off-by:

I'll leave some comments in the commits.

gkurz commented on Makefile:
```diff
@@ -312,6 +312,7 @@ bundle-clean: ## Clean generated bundle files
 
 .PHONY: bin-clean
 bin-clean: ## Clean downloaded binaries
+	@[ -d bin ] && chmod -R u+rwx bin || true
```
I don't quite understand how `chmod` would succeed in a scenario where `rm -f -r` would fail... what was the problem ?

tbuskey replied:
Permissions in bin/k8s/1.24.2-linux-amd64 were readonly:
dr-xr-xr-x. 2 tbuskey tbuskey 4096 Jul  3 13:07 1.24.2-linux-amd64

gkurz replied:
This is the result of `make test` failing and not fixing the mode as per bef333e. It means that the SKIP_TESTS approach isn't a good workaround for the broken unit tests. Let's revert 48942c6 and ignore errors at https://github.com/openshift/sandboxed-containers-operator/blob/devel/Makefile#L128 instead. I'll create a PR for this.

gkurz replied:
Now that #429 is merged, please rebase and you can drop this commit.

gkurz commented on Makefile:
```diff
@@ -257,7 +257,7 @@ ifeq (,$(shell which opm 2>/dev/null))
 	mkdir -p $(dir $(OPM)) ;\
 	OS=$(shell go env GOOS) && ARCH=$(shell go env GOARCH) && \
-	curl -sSLo $(OPM) https://github.com/operator-framework/operator-registry/releases/download/v1.23.0/$${OS}-$${ARCH}-opm ;\
+	curl -sSLo $(OPM) https://github.com/operator-framework/operator-registry/releases/download/v1.28.0/$${OS}-$${ARCH}-opm ;\
```
Please provide justification for this bump.

tbuskey replied:
opm 1.28 matches the current version of operator-sdk

- go version bumped to 1.18 -> 1.19
- k8s bumped 1.24 -> 1.25 -> 1.26
- go libraries bumped
- FBC caching refactored
- fix channel sorting
- improve selinux labeling
- bug fixes

gkurz commented on Makefile:
```diff
@@ -272,17 +272,25 @@ BUNDLE_IMGS ?= $(BUNDLE_IMG)
-# Build a catalog image by adding bundle images to an empty catalog using the operator package manager tool, 'opm'.
-# This recipe invokes 'opm' in 'semver' bundle add mode. For more information on add modes, see:
+# Build a file based catalog image using opm.
+# https://docs.openshift.com/container-platform/4.12/operators/admin/olm-managing-custom-catalogs.html
```
Please point to latest documentation at https://docs.openshift.com/container-platform/4.16/operators/admin/olm-managing-custom-catalogs.html.

gkurz commented on Makefile:
```diff
@@ -272,17 +272,25 @@ BUNDLE_IMGS ?= $(BUNDLE_IMG)
-catalog-build: opm ## Build a catalog image.
-	$(OPM) index add --container-tool docker --mode semver --tag $(CATALOG_IMG) --bundles $(BUNDLE_IMGS) $(FROM_INDEX_OPT)
+catalog-build: opm
+	@[ ! -d catalog ] && mkdir -p catalog || true
```
`mkdir -p` just exits 0 if the directory already exists.
```suggestion
	mkdir -p catalog
```

tbuskey replied:
It was exiting the make when I was testing

gkurz commented on Makefile:
```diff
@@ -272,17 +272,25 @@ BUNDLE_IMGS ?= $(BUNDLE_IMG)
-	$(OPM) index add --container-tool docker --mode semver --tag $(CATALOG_IMG) --bundles $(BUNDLE_IMGS) $(FROM_INDEX_OPT)
+catalog-build: opm
+	@[ ! -d catalog ] && mkdir -p catalog || true
+	@[ ! -f catalog.Dockerfile ] && $(OPM) generate dockerfile catalog || true
```
It looks like `catalog.Dockerfile` should be a prereq for the `catalog-build` target. This will allow to get rid of the `|| true` that would hide a failure of `$(OPM)` AFAICT.

tbuskey replied:
_opm generate dockerfile catalog_ gives an error if **catalog.Dockerfile** exists

tbuskey replied:
catalog/index.yaml and catalog.Dockerfile are now checked in
Regenerating them is now in DEVELOPMENT.md

gkurz commented:
This PR isn't ready to be merged as is. Some more polishing is needed.

- when addressing a reviewer's requests, it is preferrable to squash the fixes in the original commits and force-push as suggested at https://docs.google.com/document/d/1aZEuqUubGWassHyURs7t_4sX1fMAnxyaBAlKsS8KbvU/edit#heading=h.6nn0bdwndby1 and https://docs.google.com/document/d/1aZEuqUubGWassHyURs7t_4sX1fMAnxyaBAlKsS8KbvU/edit#heading=h.xocsdceerbc
e.g. Makefile changes from https://github.com/openshift/sandboxed-containers-operator/pull/422/commits/0a61d070f2bc2a44bede7700e21b9d6cda1e5d16 should be squashed into https://github.com/openshift/sandboxed-containers-operator/pull/422/commits/42536b539025279511170ebf70ddad8acb3aba3e
and documentation changes should be squashed into https://github.com/openshift/sandboxed-containers-operator/pull/422/commits/461059c48cb6b16cf6baf77539c08915743886ea
- all commits should have a descriptive title, a good commit message and a `Signed-off-by` tag as suggested by https://docs.google.com/document/d/1aZEuqUubGWassHyURs7t_4sX1fMAnxyaBAlKsS8KbvU/edit#heading=h.jwlp2948280n and https://docs.google.com/document/d/1aZEuqUubGWassHyURs7t_4sX1fMAnxyaBAlKsS8KbvU/edit#heading=h.iwk7eyleqzrc

tbuskey committed:
Makefile: Add the --authfile & location of the tokens for
registry.ci.openshift.org

The BUILDER_IMAGE and TARGET_IMAGE are retrieved from
registry.ci.openshift.org which requires authentication and token.  If
your token is not in ~/.docker/config.json, AUTH_FILE can be set to a
different filename.

Signed-off-by: Tom Buskey <tbuskey@redhat.com>

tbuskey committed:
Makefile: bump opm version to 1.28

opm has been updated w/ new golang versions and refactored.

Signed-off-by: Tom Buskey <tbuskey@redhat.com>

tbuskey committed:
Makefile: catalog-build is changed to a File Based Catalog

OLM (and opm) have deprecated sqlite catalogs for FBC catalogs.
  catalog-build has been changed
  catalog/index.yaml is created by opm init/opm render
  catalog.Dockerfile is created by opm generate

Signed-off-by: Tom Buskey <tbuskey@redhat.com>

tbuskey committed:
docs/DEVELOPMENT.md: Updated to match Makefile

- Explain changes in the Makefile
- Explain how to get tokens to download build and target images
  - remove public image section
- environment variables that can be set
- regenerating the catalog

Signed-off-by: Tom Buskey <tbuskey@redhat.com>

gkurz commented on catalog/index.yaml:
```diff
@@ -0,0 +1,243 @@
+name: sandboxed-containers-operator
+schema: olm.package
+---
+image: quay.io/openshift_sandboxed_containers/openshift-sandboxed-containers-operator-bundle:v1.6.0
```
```suggestion
image: quay.io/openshift_sandboxed_containers/openshift-sandboxed-containers-operator-bundle:v1.7.0
```

tbuskey replied:
Changing the version

gkurz commented on catalog/index.yaml:
```diff
@@ -0,0 +1,243 @@
+schema: olm.package
+---
+image: quay.io/openshift_sandboxed_containers/openshift-sandboxed-containers-operator-bundle:v1.6.0
+name: sandboxed-containers-operator.v1.6.0
```
```suggestion
name: sandboxed-containers-operator.v1.7.0
```

gkurz commented on catalog/index.yaml:
```diff
@@ -0,0 +1,243 @@
+- type: olm.package
+  value:
+    packageName: sandboxed-containers-operator
+    version: 1.6.0
```
```suggestion
    version: 1.7.0
```

tbuskey replied:
Changing the version

gkurz commented on catalog/index.yaml:
```diff
@@ -0,0 +1,243 @@
+      features.operators.openshift.io/token-auth-aws: "false"
+      features.operators.openshift.io/token-auth-azure: "false"
+      features.operators.openshift.io/token-auth-gcp: "false"
+      olm.skipRange: '>=1.1.0 <1.5.2'
```
```suggestion
      olm.skipRange: '>=1.1.0 <1.7.0'
```

gkurz commented on catalog/index.yaml:
```diff
@@ -0,0 +1,243 @@
+relatedImages:
+- image: gcr.io/kubebuilder/kube-rbac-proxy@sha256:d99a8d144816b951a67648c12c0b988936ccd25cf3754f3cd85ab8c01592248f
+  name: ""
+- image: quay.io/openshift_sandboxed_containers/openshift-sandboxed-containers-operator-bundle:v1.6.0
```
```suggestion
- image: quay.io/openshift_sandboxed_containers/openshift-sandboxed-containers-operator-bundle:v1.7.0
```

tbuskey replied:
Changing the version

gkurz commented on catalog/index.yaml:
```diff
@@ -0,0 +1,243 @@
+- image: quay.io/openshift_sandboxed_containers/openshift-sandboxed-containers-operator:v1.6.0
+  name: ""
+- image: registry.redhat.io/openshift-sandboxed-containers/osc-cloud-api-adaptor-rhel9:latest
+  name: caa
```
Why do previous images have `name: ""` ?

tbuskey replied:
opm generate creates the index.yaml.  It starts at 1.6.0 and not something earlier

gkurz commented on catalog/index.yaml:
```diff
@@ -0,0 +1,243 @@
+package: sandboxed-containers-operator
+name: stable
+entries:
+  - name: sandboxed-containers-operator.v1.6.0
```
```suggestion
  - name: sandboxed-containers-operator.v1.7.0
```

gkurz replied:
AFAIK we also need the following because of `olm.skipRange` :

```
    replaces: sandboxed-containers-operator.v1.6.0
```

See [KATA-2894](https://issues.redhat.com/browse/KATA-2894) for an explanation.

tbuskey replied:
Adding

gkurz reviewed (commented):
Great work @tbuskey ! It just requires to bump the OSC version to 1.7.0 in the catalog index file. I'll try it out and approve.

gkurz commented on docs/DEVELOPMENT.md:
```diff
@@ -12,82 +12,109 @@ install -m 755 operator-sdk_linux_amd64 ${SOME_DIR_IN_YOUR_PATH}/operator-sdk
+package: sandboxed-containers-operator\n\
+name: ${DEFAULT_CHANNEL}\n\
+entries:\n\
+  - name: sandboxed-containers-operator.v${VERSION}\n\
```
As explained at https://github.com/openshift/sandboxed-containers-operator/pull/422/commits/55a01ed18c6249eb9a42f1e9c02862546f2c7f1c#r1673838632, please something like :
```
    replaces: sandboxed-containers-operator.v${PREVIOUS_VERSION}\n\
```

tbuskey replied:
Done

gkurz commented on catalog/index.yaml:
```diff
@@ -0,0 +1,243 @@
+  name: ""
+- image: quay.io/openshift_sandboxed_containers/openshift-sandboxed-containers-operator-bundle:v1.6.0
+  name: ""
+- image: quay.io/openshift_sandboxed_containers/openshift-sandboxed-containers-operator:v1.6.0
```
```suggestion
- image: quay.io/openshift_sandboxed_containers/openshift-sandboxed-containers-operator:v1.7.0
```

gkurz commented on catalog/index.yaml:
```diff
@@ -0,0 +1,243 @@
+  name: ""
+- image: quay.io/openshift_sandboxed_containers/openshift-sandboxed-containers-operator:v1.6.0
+  name: ""
+- image: registry.redhat.io/openshift-sandboxed-containers/osc-cloud-api-adaptor-rhel9:latest
```
We try to avoid using the `latest` tag as explained in f8169dbedc131b60c1268c7757a0afd84523498b.

tbuskey replied:
The `latest` versioning for the podvm probably comes from` config/peerpods/podvm`.
We will have to investigate the others

tbuskey committed:
Bump versions in catalog from 1.6.0 to 1.7.0

Everything should be 1.7.0 to match the Makefile
DEVELOPMENT.md directions to generate index.yaml were also updated to
have PREVIOUS_VERSION

Signed-off-by: Tom Buskey <tbuskey@redhat.com>

gkurz commented on docs/DEVELOPMENT.md:
```diff
@@ -12,82 +12,109 @@ install -m 755 operator-sdk_linux_amd64 ${SOME_DIR_IN_YOUR_PATH}/operator-sdk
+BUNDLE_IMG=${IMAGE_TAG_BASE}-bundle:v${VERSION}
+make opm                  # to get bin/opm
+rm -rf catalog && mkdir -p catalog
+bin/opm init  sandboxed-containers-operator --default-channel=${DEFAULT_CHANNEL} --description ./README.md --output yaml > catalog/index.yaml
```
I know this was discussed previously at https://github.com/openshift/sandboxed-containers-operator/pull/422#discussion_r1661202637 but something now looks wrong to me.

This `bin/opm init ...` has a dependency on `README.md` : it should be run again every time `README.md` is changed, not just once. Typical case for a rule in the Makefile.

Also I still don't understand why `catalog/index.yaml` should be committed in the repo.

@cpmeadors ?

gkurz commented:
Converted to draft to block merge until this PR is fully reviewed.

bpradipt commented:
@tbuskey since we rely on operator-sdk, one suggestion is to create a skeleton Makefile using operator-sdk and align your changes to that. Example you can following the instructions [here](https://sdk.operatorframework.io/docs/building-operators/golang/quickstart/#steps) to create a skeleton Makefile and compare it with what we have and add your changes.

For example, I see that latest Makefile has options to download all the required binaries including operator-sdk locally. Let's follow that and ignore my previous comment on (not) downloading the operator-sdk. It's better to align with operator-sdk generated skeleton as it helps when we migrate to newer versions of sdk. We already are facing issues with deviating from the sdk generated skeleton file `.gitignore`, so let's avoid deviating from the sdk generated skeleton as reference unless we have a really really strong reason.
Also looking at other operators and how they are building file based catalog will provide insights on what is needed.
Example - https://github.com/openshift/lvm-operator

gkurz commented on docs/DEVELOPMENT.md:
````diff
@@ -12,82 +12,109 @@ install -m 755 operator-sdk_linux_amd64 ${SOME_DIR_IN_YOUR_PATH}/operator-sdk
+#### Using the token
+If docker build does not automatically use the token in the default authfile, it can be forces:
+```shell
+export AUTH_FILE='~/.docker/config.json'
````
From the `podman-login(1)` manual page :
```
The default path for reading and writing credentials is ${XDG_RUNTIME_DIR}/containers/auth.json.
Podman uses existing credentials if the user does not pass in a username.  Podman first searches
for  the  username and  password  in  the ${XDG_RUNTIME_DIR}/containers/auth.json,  if  they are
not valid, Podman then uses any existing credentials found in $HOME/.docker/config.json.
```

```suggestion
export AUTH_FILE="${XDG_RUNTIME_DIR}/containers/auth.json"
```

gkurz replied:
Hmm... I gave a try without setting `AUTH_FILE` and it worked :
- `podman login` does add the credentials to `"${XDG_RUNTIME_DIR}/containers/auth.json"` 
- `podman build` does find the credentials there
I now suspect that `AUTH_FILE` isn't needed at all.

What were the steps that ended up in `docker build does not automatically use the token in the default authfile` ?
